audio/out/buffer.c

```diff
@@ -117,8 +126,29 @@ struct mp_async_queue *ao_get_queue(struct ao *ao)
     return p->queue;
 }
 
+// called locked
+static void prune_queued_sameples(struct ao *ao)
+{
+    struct buffer_state *p = ao->buffer_state;
+    int pos;
+    double cur_time = MP_TIME_NS_TO_S(mp_time_ns());
+
+    for (pos = 0; pos <= p->num_queued_aframes; pos++) {
+        if (pos == p->num_queued_aframes || p->queued_aframes[pos].start_time > cur_time)
+            break;
+    }
```

Comment:
wouldn't this be the same?
```
    for (pos = 0; pos < p->num_queued_aframes; pos++) {
        if (p->queued_aframes[pos].start_time > cur_time)
            break;
    }
```

Reply:
My mistake. Should be:

```C
    for (pos = 0; pos <= p->num_queued_aframes; pos++) {
        if (pos < p->num_queued_aframes && p->queued_aframes[pos].start_time > cur_time)
            break;
    }
```

`queued_aframes` from `0` to `pos - 2` are removed. By using `pos <= p->num_queued_aframes`, `pos` can be `p->num_queued_aframes + 1` if the `start_time` of all aframes are greater than `cur_time`. `p->num_queued_aframes + 1` minus `2` equal to `p->num_queued_aframes - 1`, hence all aframes can be pruned. If we use `pos < p->num_queued_aframes`, at least one aframe is always left.